### 파이썬 DB 연동
- CRUD 작업 처리
- CREATE / READ / UPDATE / DELETE
- INSERT / SELECT / UPDATE / DELETE 

### 파이썬에서 MySQL 연동 패키지 지원
- pymysql 패키지 사용 

In [1]:
# pymysql 패키지 설치 
!pip install pymysql

In [2]:
import pymysql

#### DB 연동 순서
- (1) DB 연결   
- (2) cursor 객체 생성 
- (3) sql 구문 실행 
- (4) sql 실행 결과 추출 (출력)

In [3]:
# 1. DB 연결 
# 호스트명, 포트번호, 로그인 계정, 비밀번호, CHARSET 파라미터로 지정 
conn = pymysql.connect(host='localhost', 
                       port=3306, 
                       user='root', 
                       passwd='1234',
                       charset='utf8')
# 연결 객체 반환 

In [4]:
# 2. cursor 객체 생성 
# cursor 객체를 통해서 메소드 호출 
cursor = conn.cursor()

In [5]:
# 3. sql 문 실행 
# 데이터베이스 확인 sql 문 실행 
# 실행 : execute() 메소드 
sql = 'show databases'
cursor.execute(sql)

# 실행 결과 : 데이터베이스 개수 반환 

11

In [6]:
# 4. sql 실행 결과 추출 (출력)
# 서버로부터 받은 결과를 출력 
# fetchall() 메소드 사용 : 모든 데이터를 한꺼번에 클라이언트로 받아올 때 사용 
result = cursor.fetchall()
result

(('information_schema',),
 ('mysql',),
 ('performance_schema',),
 ('sakila',),
 ('shopdb',),
 ('sqldb1',),
 ('sqldb2',),
 ('sqldb3',),
 ('sys',),
 ('testdb',),
 ('world',))

In [7]:
# !pip install cryptography

In [8]:
# 새로운 데이터베이스 생성 
sql = 'create database beauty_shop'
cursor.execute(sql)

1

In [9]:
# beauty_shop 데이터베이스 생성 확인
sql = "show databases"
cursor.execute(sql)
result = cursor.fetchall()
result

(('beauty_shop',),
 ('information_schema',),
 ('mysql',),
 ('performance_schema',),
 ('sakila',),
 ('shopdb',),
 ('sqldb1',),
 ('sqldb2',),
 ('sqldb3',),
 ('sys',),
 ('testdb',),
 ('world',))

In [10]:
# 데이터베이스 삭제 
sql = 'drop database beauty_shop'
cursor.execute(sql)

0

In [11]:
# beauty_shop 데이터베이스 삭제 확인
sql = "show databases"
cursor.execute(sql)
result = cursor.fetchall()
result

(('information_schema',),
 ('mysql',),
 ('performance_schema',),
 ('sakila',),
 ('shopdb',),
 ('sqldb1',),
 ('sqldb2',),
 ('sqldb3',),
 ('sys',),
 ('testdb',),
 ('world',))

In [13]:
# 사용할 데이터베이스 선택 
sql = 'use sqldb3'
cursor.execute(sql)

0

In [14]:
# 테이블 생성 
sql = """
    create table shop(
        proNo int auto_increment not null primary key,
        prdName varchar(100),
        prdPrice float,
        prdDisPrice float
    )
"""

cursor.execute(sql)

0

In [15]:
# 생성된 테이블 확인 
sql = "show tables"
cursor.execute(sql)
cursor.fetchall()

(('book',),
 ('booksale',),
 ('client',),
 ('client_view',),
 ('flower',),
 ('movie',),
 ('new_book1',),
 ('publisher',),
 ('shop',))

In [16]:
# 테이블 삭제 
sql = 'drop table shop'
cursor.execute(sql)

0

In [17]:
# 테이블 구조 확인
sql = 'desc book'
cursor.execute(sql)
cursor.fetchall()
# 열이름, 데이터타입(크기), null 여부, 키, 디폴트값, 기타

(('bookNo', 'varchar(10)', 'NO', 'PRI', None, ''),
 ('bookName', 'text', 'YES', '', None, ''),
 ('bookAuthor', 'text', 'YES', '', None, ''),
 ('bookPrice', 'int', 'YES', '', None, ''),
 ('bookDate', 'date', 'YES', '', None, ''),
 ('bookStock', 'int', 'YES', '', None, ''),
 ('pubNo', 'int', 'YES', '', None, ''))

In [18]:
###################################################################

In [19]:
# CRUD 작업 처리
# CREATE / READ / UPDATE / DELETE
# INSERT / SELECT / UPDATE / DELETE

In [20]:
# SELECT 

In [21]:
# book 테이블 데이터 조회 
sql = 'select * from book'
cursor.execute(sql)
result = cursor.fetchall()
result

(('1001', '데이터베이스 이론', '홍길동', 22000, datetime.date(2019, 11, 11), 5, 3),
 ('1002', '자바 프로그래밍', '이몽룡', 25000, datetime.date(2021, 12, 12), 4, 1),
 ('1003', '인터넷프로그래밍', '성춘향', 30000, datetime.date(2019, 2, 10), 10, 2),
 ('1004', '안드로이드 프로그래밍', '변학도', 23000, datetime.date(2017, 10, 10), 2, 1),
 ('1005', '안드로이드 앱', '강길동', 26000, datetime.date(2020, 1, 11), 5, 2),
 ('1006',
  'MS SQL SERVER 2014',
  '박지성',
  35000,
  datetime.date(2020, 3, 25),
  7,
  3),
 ('1007', 'HTML & CSS', '손흥민', 18000, datetime.date(2023, 9, 30), 3, 1),
 ('1008', 'MFC 입문', '류현진', 20000, datetime.date(2015, 12, 12), 5, 1),
 ('1009', '안드로이드 게임 제작', '나길동', 33000, datetime.date(2018, 10, 31), 5, 2),
 ('1010', 'C++객체지향 프로그래밍', '추신수', 24000, datetime.date(2019, 4, 20), 2, 3),
 ('1011', 'JSP 웹 프로그래밍', '김연아', 27000, datetime.date(2020, 10, 23), 8, 1),
 ('1012', '해커들의 해킹 기법', '손연재', 32000, datetime.date(2018, 7, 7), 1, 2),
 ('1013', '자료구조', '홍길동', 19000, datetime.date(2020, 1, 20), 4, 1),
 ('1014', '파이썬', '성춘향', 35000, dateti

In [22]:
result[0][0]

'1001'

In [23]:
for re in result:
    print(re)

('1001', '데이터베이스 이론', '홍길동', 22000, datetime.date(2019, 11, 11), 5, 3)
('1002', '자바 프로그래밍', '이몽룡', 25000, datetime.date(2021, 12, 12), 4, 1)
('1003', '인터넷프로그래밍', '성춘향', 30000, datetime.date(2019, 2, 10), 10, 2)
('1004', '안드로이드 프로그래밍', '변학도', 23000, datetime.date(2017, 10, 10), 2, 1)
('1005', '안드로이드 앱', '강길동', 26000, datetime.date(2020, 1, 11), 5, 2)
('1006', 'MS SQL SERVER 2014', '박지성', 35000, datetime.date(2020, 3, 25), 7, 3)
('1007', 'HTML & CSS', '손흥민', 18000, datetime.date(2023, 9, 30), 3, 1)
('1008', 'MFC 입문', '류현진', 20000, datetime.date(2015, 12, 12), 5, 1)
('1009', '안드로이드 게임 제작', '나길동', 33000, datetime.date(2018, 10, 31), 5, 2)
('1010', 'C++객체지향 프로그래밍', '추신수', 24000, datetime.date(2019, 4, 20), 2, 3)
('1011', 'JSP 웹 프로그래밍', '김연아', 27000, datetime.date(2020, 10, 23), 8, 1)
('1012', '해커들의 해킹 기법', '손연재', 32000, datetime.date(2018, 7, 7), 1, 2)
('1013', '자료구조', '홍길동', 19000, datetime.date(2020, 1, 20), 4, 1)
('1014', '파이썬', '성춘향', 35000, datetime.date(2023, 9, 25), 5, 2)
('1015', '자

In [24]:
# 도서번호, 도서명, 도서가격만 쿼리 실행 
sql = 'select bookNo, bookName, bookPrice from book';
cursor.execute(sql)
result = cursor.fetchall()
print(result)


(('1001', '데이터베이스 이론', 22000), ('1002', '자바 프로그래밍', 25000), ('1003', '인터넷프로그래밍', 30000), ('1004', '안드로이드 프로그래밍', 23000), ('1005', '안드로이드 앱', 26000), ('1006', 'MS SQL SERVER 2014', 35000), ('1007', 'HTML & CSS', 18000), ('1008', 'MFC 입문', 20000), ('1009', '안드로이드 게임 제작', 33000), ('1010', 'C++객체지향 프로그래밍', 24000), ('1011', 'JSP 웹 프로그래밍', 27000), ('1012', '해커들의 해킹 기법', 32000), ('1013', '자료구조', 19000), ('1014', '파이썬', 35000), ('1015', '자바스크립 & jQuery', 27000))


In [25]:
# 도서 가격이 30,000원 이상인 도서의 도서번호, 도서명, 가격 추출 
sql = 'select bookno,bookname,bookprice from book where bookprice >=30000'

cursor.execute(sql)
result = cursor.fetchall()
result


(('1003', '인터넷프로그래밍', 30000),
 ('1006', 'MS SQL SERVER 2014', 35000),
 ('1009', '안드로이드 게임 제작', 33000),
 ('1012', '해커들의 해킹 기법', 32000),
 ('1014', '파이썬', 35000))

# 다음과 같이 출력  
1003 인터넷프로그래밍 30000
1006 MS SQL SERVER 2014 35000
1009 안드로이드 게임 제작 33000
1012 해커들의 해킹 기법 32000
1014 파이썬 35000

In [30]:
for re in result:
    for r in re:
        print(r, end=" ")
    print()

1003 인터넷프로그래밍 30000 
1006 MS SQL SERVER 2014 35000 
1009 안드로이드 게임 제작 33000 
1012 해커들의 해킹 기법 32000 
1014 파이썬 35000 


In [31]:
cursor.close()
conn.close()